apps/emqx_auth/src/emqx_authz/emqx_authz_schema.erl

```diff
@@ -162,10 +170,23 @@ api_authz_fields() ->
     [{sources, ?HOCON(?ARRAY(api_source_type()), #{desc => ?DESC(sources)})}].
 
 api_source_type() ->
-    hoconsc:union(api_authz_refs()).
-
-api_authz_refs() ->
-    lists:concat([api_source_refs(Mod) || Mod <- source_schema_mods()]).
+    AuthzSchemaMods = source_schema_mods(),
+    AllTypes = lists:concat([api_source_refs(Mod) || Mod <- AuthzSchemaMods]),
+    UnionMemberSelector =
+        fun
+            (all_union_members) ->
+                AllTypes;
+            %% must return list
+            ({value, Value}) ->
+                select_union_member(
+                    Value,
+                    AuthzSchemaMods,
+                    fun(Mod, _) ->
+                        api_source_refs(Mod)
```

Comment:
should fix emqx_authz_file_schema:select_union_member/2 to select type based on path or rules. ?